Initial code

rfairley opened the pull request.

Commit:
Working main.rs

Commit:
main: add comment for run()

Commit:
systemd: fix service description

Commit:
add default config file

Commit:
add .travis.yml

Commit:
Cargo.toml: add info for packaging

Commit:
add specfile

Commit:
README: add dependency note

rfairley commented on src/main.rs:
```diff
@@ -0,0 +1,46 @@
+#[macro_use]
+extern crate log;
+
+use glib::{KeyFile, KeyFileFlags};
```
Not attached to the GLib keyfile format here (https://developer.gnome.org/glib/stable/glib-Key-value-file-parser.html#glib-Key-value-file-parser.description) - could switch this to use a TOML format for the `client.conf` file instead (either in this PR or later).

lucab replied:
I think it's indeed a better idea to just use TOML since the beginning.

rfairley replied:
:+1: - fixed.

lucab commented on src/main.rs:
```diff
@@ -0,0 +1,46 @@
+
+use glib::{KeyFile, KeyFileFlags};
+
+const CONFIG_FILE_PATH: &str = "/etc/fedora-coreos-metrics/client.conf";
```
I'd suggest to just reserve `/{run,etc,usr/lib}/fedora-coreos-metrics/config.d/` since the beginning and pick up config files from there.

lucab commented on dist/client.conf:
```diff
@@ -0,0 +1,5 @@
+
+[metrics]
+# Default metrics.level is `minimal`. May be set to `minimal`, `full`, or `off`.
+level=minimal
```
I think you may want to split this knob into two: a "collecting" section with "level" (minimal/full/etc), and a "reporting" section with "enabled" (bool).

rfairley replied:
:+1: sounds better to separate them - added this change!

Commit:
main: switch to toml for config parsing

Commit:
separate client.conf fields

Separate the fields into a collection.level and reporting.enabled
rather than one field metrics.level.

Commit:
main: remove InvalidMetricsLevel error

Using bail!() is sufficient for now.

Commit:
main: reorder quick_main! and update comment

Commit:
README: update and remove glib2-devel note

Commit:
parse config snippets

Adapts code providing similar functionality for coreos/zincati:
https://github.com/coreos/zincati/blob/60f3a9144b34ebfa7f7a0fe98f8d641a760ee8f0/src/config/inputs.rs.

Also use `failure` instead of `error_chain`.

Commit:
main: factor into helper function

Commit:
regenerate specfile from crate

Commit:
specfile: install systemd unit and dirs

lucab commented on src/main.rs:
```diff
@@ -0,0 +1,152 @@
+            debug!("scanning configuration directory '{}'", dir);
+
+            let wildcard = format!("{}/*.toml", dir);
+            let toml_files = glob::glob(&wildcard)?;
```
I think I've seen this code already ;)
Please do note that I quickly put this together for zincati in order to use snippets from multiple directories, but it doesn't understand same-name overrides.

rfairley replied:
Yes - shamelessly copied :) Thanks - yes I'd like to support same-name overrides here - pushed support for this (still need to add a test, and it likely can be optimized to not store the fragments in the BTreeMap).

lucab replied:
I think you could store either a `PathBuf` or a `File`, and let consumers later transform those into fragments (later reduced into a single config).

lucab replied:
Similar logic is at https://github.com/overdrop/overdrop-sebool/blob/master/src/od_cfg.rs#L26.

Commit:
main: support config overlays

Commit:
main: store PathBuf in config BTreeMap

Commit:
move systemd/ -> dist/systemd/

Commit:
main: remove Fallible from use directives

Commit:
Cargo: allow semver updates

Commit:
main: general refactoring

Commit:
main: split out input and fragments modules

Commit:
split modules into files

Commit:
add basic config test

jlebon commented on README.md:
```diff
@@ -1 +1,13 @@
-# fedora-coreos-metrics-client
\ No newline at end of file
+# fedora-coreos-metrics-client
+
```
Would be good to have a short description here of what the goal of this project is.

rfairley replied:
Thanks, added this! Also just added in some documentation on configuring it.

Commit:
README: add note on goal and configuration

Commit:
specfile: fix config fragment path

Commit:
rename default client config file

Commit:
config: add a few docstrings

rfairley commented:
OK - tidied up a few things! Plan to merge by end of tomorrow.

Thank you for the feedback!
